build_tools/benchmarks/collect_compilation_statistics.py

```diff
@@ -0,0 +1,208 @@
+#!/usr/bin/env python3
+# Copyright 2022 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+"""Collect compilation statistics from benchmark suites.
+
+The benchmark suites need to be built with ninja and enable the CMake option
+IREE_ENABLE_COMPILATION_BENCHMARKS.
+"""
+
+import argparse
+import json
+import os
+import re
+import zipfile
+
+from dataclasses import asdict
+from typing import BinaryIO, Dict, Optional, TextIO
+from pathlib import PurePath
+
+from common.benchmark_definition import CompilationInfo, CompilationResults, CompilationStatistics, ModuleComponentSizes, get_git_commit_hash
+from common.benchmark_suite import BENCHMARK_SUITE_REL_PATH, BenchmarkSuite
+
+BENCHMARK_FLAGFILE = "flagfile"
+MODULE_DIR = "vmfb"
+MODULE_FILE_EXTENSION = ".vmfb"
+NINJA_LOG_HEADER = "ninja log v5"
+NINJA_BUILD_LOG = ".ninja_log"
+COMPILATION_STATS_MODULE_SUFFIX = "compile-stats"
+
+VM_COMPONENT_NAME = "module.fb"
+CONST_COMPONENT_NAME = "_const.bin"
+DISPATCH_COMPONENT_PATTERNS = [
+    r".+_embedded_elf_.+\.so",
+    r".+_vulkan_spirv_fb\.fb",
+    r".+_vmvx_bytecode_fb\.bin",
+]
+
+
+def match_module_cmake_target(module_path: str) -> Optional[str]:
+  # Get the last 4 parts of module path. They are expected to be:
+  # benchmark_suites/<category>/vmfb/<module filename>.vmfb
+  path_parts = PurePath(module_path).parts[-4:]
+  if len(path_parts) < 4:
+    return None
+  if path_parts[0] != BENCHMARK_SUITE_REL_PATH:
+    return None
+  if path_parts[2] != MODULE_DIR:
+    return None
+  if os.path.splitext(path_parts[3])[1] != MODULE_FILE_EXTENSION:
+    return None
+  return os.path.join(*path_parts)
+
+
+def parse_compilation_time_from_ninja_log(log: TextIO) -> Dict[str, int]:
+  """Retrieve the compilation time (ms) from the Ninja build log.
+  
+  Returns:
+    Map of target name and compilation time in ms.
+  """
+
+  target_build_time_map = {}
+  header = log.readline()
+  if NINJA_LOG_HEADER not in header:
+    raise NotImplementedError(f"Unsupported ninja log version: {header}")
+
+  for line in log:
+    start_time, end_time, _, target, _ = line.strip().split("\t")
+    cmake_target = match_module_cmake_target(target)
+    if cmake_target is None:
+      continue
+
+    start_time = int(start_time)
+    end_time = int(end_time)
+    target_build_time_map[cmake_target] = end_time - start_time
+
+  return target_build_time_map
+
+
+def get_module_component_info(module: BinaryIO,
+                              module_file_size: int) -> ModuleComponentSizes:
+  with zipfile.ZipFile(module) as module_zipfile:
+    size_map = dict(
+        (info.filename, info.file_size) for info in module_zipfile.infolist())
```

Comment:
The file size of binaries often has surprisingly coarse granularity and tracking just that may fail to observe incremental code size increase, until some threshold is reached and the file size finally bumps.  In my experience (and that is what we have converged on in internal google CI for code size regression tests), something that works reasonably well is `size -A <thefile>`, and then look at the last line, `Total`. Here is an example. Consider these two small C++ shared libraries, `foo` and `bar`:

```shell
benoitjacob@desk:~$ cat foo.cc 
#include <unordered_map>

int foo(int a, int b) {
  std::unordered_map<int, int> m;
  m[a] = b;
  m[b] = a;
  return m[a] + m[b];
}

benoitjacob@desk:~$ cat bar.cc
#include <unordered_map>

int foo(int a, int b) {
  std::unordered_map<int, int> m;
  m[a] = b;
  m[b] = a;
  m[a + b] = a - b;
  m[a - b] = a + b;
  if (m.count(123)) return 456;
  return m[a] + m[b];
}

benoitjacob@desk:~$ clang++ foo.cc -g0 -O2 -shared -fPIC -o /tmp/foo.so
benoitjacob@desk:~$ clang++ bar.cc -g0 -O2 -shared -fPIC -o /tmp/bar.so
benoitjacob@desk:~$ ls -l /tmp/{foo,bar}.so
-rwxr-xr-x 1 benoitjacob primarygroup 17232 Jun 10 10:11 /tmp/bar.so
-rwxr-xr-x 1 benoitjacob primarygroup 17232 Jun 10 10:11 /tmp/foo.so
```

So both libraries, `foo.so` and `bar.so`, have the same file size, 17232 bytes, despite `bar` doing a bit more work.

Fortunately, `size` can see the difference:

```shell
benoitjacob@desk:~$ size -A /tmp/foo.so | grep Total
Total                5966
benoitjacob@desk:~$ size -A /tmp/bar.so | grep Total
Total                6602
```

This is a > 10% code size increase!

Reply:
Thanks! @bjacob , sorry I didn't pay enough attention to it.

I was thinking that we might also have similar issues on other payloads (E.g., SPIR-V framebuffer), so I preferred not to do ad-hoc things for dynamic libraries first. I'll spend some time checking the better ways to do this #9436.

Just tried to think about the cause of the difference, I guess that might be due to the alignment of the ELF section, so each section is padded to the page size.

Reply:
:+1:
agree it looks like some kind of alignment. but in my experience, it's often more than page size.  i seem to remember it was 64k in internal google android aarch64 builds of TFlite.